[WIP] Discover - allow "Time" column to behave like a regular column

kobelb opened the pull request:
This is a work-in-process, and existing saved searches will not play nicely with how this is currently implemented.

The CSS was hard-coded in some situations with the anticipation that the Time Field would be in the first column a majority of times, so this had to be refactored slightly. There's an existing UI/UX glitch, where expanding the row in Discover can sometimes cause the columns size to jump around . I'm waiting on investigating this further until we determine whether this approach is how we want to proceed, and whether the migrations will allow us to make it work with existing saved searches.

Intended to address https://github.com/elastic/kibana/issues/3319

lukasolson commented:
@weltenwort has been working on refactoring the doc table, probably a good idea for you two to sync on this one before getting too far into it :)

weltenwort commented:
This is totally in line with what I was planning too, so thanks for tackling it. No reservations from my side. :+1:

weltenwort commented:
Let me know when you think this is ready for a detailed review.

kobelb commented:
Alrighty, I've resolved the columns jumping around by hard-coding the width for the expand column based off the open column width: https://github.com/elastic/kibana/pull/11504/files#diff-83d36a35a8ff80c1778875629bf83081

Commit https://github.com/elastic/kibana/pull/11504/commits/038f6cae28687cfdbb2aa215cff77b0174e3f9c5 also includes a possible migration strategy if the migration tool isn't ready for 5.5, but hopefully this is something we can remove. I'll know more about this Monday.

kobelb commented:
@weltenwort if you're available to give this a more detailed review, it'd be much appreciated!

weltenwort commented on src/core_plugins/kibana/public/discover/controllers/discover.js:
```diff
@@ -438,6 +451,13 @@ function discoverController($scope, config, courier, $route, $window, Notifier,
+          // since flattenHit removed the _source field, but we want user's to be able to select it,
+          // we're manually adding it back
+          if (hit._source) {
+            fields.push('_source');
```
Even though the `_source` field is now present in the field chooser list, it can not be added, because the "add" button is hidden [in the template](https://github.com/kobelb/kibana/blob/038f6cae28687cfdbb2aa215cff77b0174e3f9c5/src/core_plugins/kibana/public/discover/components/field_chooser/discover_field.html#L17).

kobelb replied:
Good catch.

weltenwort commented on src/core_plugins/kibana/public/discover/controllers/discover.js:
```diff
@@ -147,6 +147,12 @@ function discoverController($scope, config, courier, $route, $window, Notifier,
 
+  // we used to by default always add in the timeFieldName implicitly, so we're
+  // automatically adding this in for the user for legacy savedSearches
+  if (!savedSearch.explicitTimeColumn && $scope.indexPattern.timeFieldName) {
```
`!savedSearch.explicitTimeColumn` is true for new saved searches as well, which means they never get the default columns further down: `savedSearch.columns.length > 0 ? savedSearch.columns : defaultColumns`. That is probably not intended, is it?

kobelb replied:
Nope, dunno how I missed that...

weltenwort commented on src/core_plugins/kibana/public/discover/controllers/discover.js:
```diff
@@ -159,10 +165,16 @@ function discoverController($scope, config, courier, $route, $window, Notifier,
+
+    const timeFieldName = $scope.indexPattern.timeFieldName;
+    if (timeFieldName) {
+      defaultColumns.unshift(timeFieldName);
```
What if `defaultColumns` already contains the time field?

kobelb replied:
This brings up an interesting situation. Currently, if the user was to set the `defaultColumns` advanced setting to "_source, @timestamp (or the equivalent timestamp name)" the Discover app would display the following three columns: Time, @timestamp, _source.

The way that I've implemented this change mirrors this behavior, and even though we're by default adding the `indexPattern.timeFieldName` to the list of columns, it won't by default be listed in the advanced settings "defaultColumns" setting. Likewise, if the user were to add the "@timestamp" to the list of defaultColumns in the advanced settings, it would show up twice in the Discover App.

We could potentially not add the `timeFieldName` to the list of defaultColumns if it's already specified there? Would that make this behavior any more user friendly?

weltenwort replied:
Right, so this is no worse than the current behavior. And having `@timestamp` in the `defaultColumns` would make it show up in non-timestamped index patterns as well, which is no good.

weltenwort commented on src/ui/public/doc_table/components/table_row/cell.html:
```diff
@@ -1,7 +1,7 @@
 var attributes = '';
 if (timefield) {
-  attributes='class="discover-table-timefield" width="1%"';
+  attributes='class="discover-table-timefield"';
```
I wish we could get rid of the special treatment here as well. This class seems to just set the `white-space: nowrap` style - which would make sense on other date columns as well. Unfortunately, a simple solution does not come to mind right now... :thinking: it bears some resemblance with https://github.com/kobelb/kibana/blob/038f6cae28687cfdbb2aa215cff77b0174e3f9c5/src/ui/public/doc_table/components/table_row.js#L180, even though that is concerned with the height.

kobelb replied:
I completely agree but failed to come up with a simple solution as well.

weltenwort commented on src/ui/public/doc_table/doc_table.html:
```diff
@@ -104,6 +104,11 @@
 </div>
 
+<div ng-if="!columns.length" class="table-vis-error">
+  <h2><i class="fa fa-meh-o"></i></h2>
```
I remember @cjcenizal telling me to not use `<i>` for the icons, but a `span` with the `kuiIcon` class...

kobelb replied:
I naively copied the markup for the "No results found" message, do you think it's worthwhile to make the new message use the better markup suggested by @cjcenizal in this PR, and deviate from the message directly below?

weltenwort replied:
How about also changing that message while you're at it? But maybe double-check with @cjcenizal just in case I misunderstood something.

cjcenizal replied:
I think it would be good to clean this up while we're touching it. At least that's what I try to do as long as the changes aren't too invasive. So we would change this to:

```html
<div ng-if="!columns.length" class="table-vis-error">
  <p class="kuiSubTitle kuiVerticalRhythm">
    <span class="kuiIcon fa-meh-o"></span>
  </p>
  
  <p class="kuiText kuiVerticalRhythm">
    No columns selected. (or whatever you think is best here)
  </p>
</div>
```

weltenwort commented on src/ui/public/doc_table/doc_table.js:
```diff
@@ -66,23 +66,6 @@ uiModules.get('kibana')
         $scope.limit += 50;
       };
 
-      // This exists to fix the problem of an empty initial column list not playing nice with watchCollection.
```
good riddance :smile:

kobelb commented:
I believe that I've addressed the issues with the explicitTimeColumn for Discover, without using the migrations; however, there's now an issue with Dashboards.

Dashboards store their own list of columns for saved searches, so that user's can re-order/add/remove columns without them affecting the underlying saved-search; which would also need to be modified to support the migration-less approach.

It's hit that point where migrations might become a requirements, as I'm unsure whether we'd like to accumulate the technical debt of the solutions working around this.

kobelb commented:
Closing this for now as road-blocks were hit and this was essentially abandoned.

alexfrancoeur commented:
@kobelb could we re-open this? We have customers that would still like to see timestamps behave as any other column in discover. 

It looks like this PR may have been closed due to this comment

> Dashboards store their own list of columns for saved searches, so that user's can re-order/add/remove columns without them affecting the underlying saved-search; which would also need to be modified to support the migration-less approach.

Given @Stacey-Gammon's work on https://github.com/elastic/kibana/pull/14452, is that still a requirement/concern? If this is not the case, what other road-blocks are there?

kobelb commented:
@alexfrancoeur there's no technical reason why this can't be re-opened. I initially tried to solve this on a Fix-It-Friday but it quickly turned out to be more complicated than I hoped. Given what we've found, I'd recommend adding it to the roadmap for either @elastic/kibana-discovery  or @elastic/kibana-sharing  and prioritizing it with the other items.

alexfrancoeur commented:
@kobelb sounds like a Discover item to me. @AlonaNadler @Bargs @lukasolson - thoughts?

AlonaNadler commented:
We have this request in https://github.com/elastic/kibana/issues/3319,  @Bargs @lukasolson  this issue has a lot of comments, can it fit in our low fruit roadmap or is it more complicated?

Bargs commented:
The biggest issue we ran into had to do with migrations of the saved search objects. I don't remember all the details atm but this is definitely more complicated than it looks.
